File: lib-segment/src/main/java/websiteschema/mpsegment/tools/accurary/AbstractErrorAnalyzer.java
package websiteschema.mpsegment.tools.accurary;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

abstract class AbstractErrorAnalyzer implements ErrorAnalyzer {

    private int occurTimes = 0;
    private Map<String, Integer> wordsWithError = new ConcurrentHashMap<String, Integer>();

    @Override
    public int getErrorOccurTimes() {
        return occurTimes;
    }

    @Override
    public Map<String,Integer> getWords() {
        return sort();
    }

    @Override
    public void postAnalysis(Map<String, Integer> allWordsAndFreqInCorpus) {
    }

    protected void increaseOccur() {
        occurTimes++;
    }

    protected void addErrorWord(String word) {
        int occurs = wordsWithError.containsKey(word) ? wordsWithError.get(word) + 1 : 1;
        wordsWithError.put(word, occurs);
    }

    protected void removeErrorWord(String word) {
        wordsWithError.remove(word);
    }

    private Map<String,Integer> sort() {
        List<Map.Entry<String,Integer>> entries = new ArrayList<Map.Entry<String,Integer>>(wordsWithError.entrySet());
        Collections.sort(entries, new Comparator<Map.Entry<String,Integer>>() {
            @Override
            public int compare(Map.Entry<String,Integer> o1, Map.Entry<String,Integer> o2) {
                return o2.getValue().compareTo(o1.getValue());
            }
        });
        Map<String, Integer> sortedMap = new LinkedHashMap<String, Integer>(entries.size());
        for(Map.Entry<String, Integer> mapEntry : entries) {
            sortedMap.put(mapEntry.getKey(), mapEntry.getValue());
        }
        return sortedMap;
    }
}

File: lib-segment/src/main/scala/websiteschema/mpsegment/tools/accurary/AbstractErrorAnalyzer.scala
package websiteschema.mpsegment.tools.accurary

import collection.mutable._
import collection.SortedMap

abstract class AbstractErrorAnalyzer extends ErrorAnalyzer {

  private var occurTimes = 0
  private var wordsWithError: Map[String, Int] = HashMap[String, Int]()

  override def getErrorOccurTimes(): Int = {
    return occurTimes
  }

  override def getWords(): ListMap[String, Int] = {
    return sort()
  }

  override def postAnalysis(allWordsAndFreqInCorpus: Map[String, Int]) {
  }

  def increaseOccur() {
    occurTimes += 1
  }

  def addErrorWord(word: String) {
    val occurs = if(wordsWithError.contains(word)) wordsWithError(word) + 1 else 1
    wordsWithError += (word -> occurs)
  }

  def removeErrorWord(word: String) {
    wordsWithError.remove(word)
  }

  private def sort(): ListMap[String, Int] = {
    return ListMap(wordsWithError.toList.sortBy{_._2}:_*)
  }
}
